[flang] Set fast math related function attributes for -Ofast/-ffast-math

asb opened the pull request:
The implemented logic matches the logic used for Clang in emitting these attributes. Although it's hoped that function attributes won't be needed in the future (vs using fast math flags in individual IR instructions), there are codegen differences currently with/without these attributes, as can be seen in issues like #79257 or by hacking Clang to avoid producing these attributes and observing codegen changes.

kiranchandramohan commented:
Is this patch addressing the issue in https://github.com/llvm/llvm-project/issues/79257 ? If not, can the other fast-math function attributes ("no-nans-fp-math"="true" and "no-signed-zeros-fp-math"="true") also be added?

asb commented:
> Is this patch addressing the issue in #79257 ? If not, can the other fast-math function attributes ("no-nans-fp-math"="true" and "no-signed-zeros-fp-math"="true") also be added?

I think we might have run into something similar to that issue independently - so thanks for pointing to that issue! I was actually checkpointing work to look more deeply into the issue as I'd heard the suspicion of differing optimisations due to these flags but didn't have a good example. I will indeed update this shortly with something more complete, and separate out the MLIR change.

asb committed:
[flang] Set fast math related function attributes for -Ofast/-ffast-math

The implemented logic matches the logic used for Clang in emitting these
attributes. Although it's hoped that function attributes won't be needed
in the future (vs using fast math flags in individual IR instructions),
there are codegen differences currently with/without these attributes,
as can be seen in issues like #79257 or by hacking Clang to avoid
producing these attributes and observing codegen changes.

asb commented:
Now ready for review - and sets all the fast-math related attributes using the same logic as Clang. The MLIR side is split out at #79812. I'm not as familiar with flang reviewers as other parts of LLVM, so please add yourself / add others as appropriate.

tblah reviewed (approved):
Thanks for working on this! LGTM

vzakhari commented on flang/lib/Optimizer/Transforms/FunctionAttr.cpp:
```diff
@@ -45,14 +50,32 @@ void FunctionAttrPass::runOnOperation() {
                                        context, framePointerKind));
 
+  if (noInfsFPMath)
+    func->setAttr("no_infs_fp_math", mlir::BoolAttr::get(context, true));
```
I am not a fan of using literal constants while they have to be kept in sync between multiple components.  Is it reasonable to make them named attributes in LLVM dialect and provide the name getters for them in `LLVMDialect.td`?

gysit replied:
Adding a function to get the attribute name from LLVMDialect.td is also not completely save since the attribute name on LLVMFuncOp is generated based on the ODS operation definition. That means things could still get out of sync.

Could it potentially make sense to run this pass after lowering to LLVMFuncOp? It would have setters for the attributes. Alternatively, LLVMFuncOp has getters to obtain the attribute names.

asb replied:
I'm happy to go with whatever is preferred here, but I agree with @gysit that it's not clear adding such getters would be a big win. We're emitting the attribute once in our code base, and if it was typoed we'd see it fail to show up (due to being unrecognised) in the test case I added.

vzakhari replied:
My main concern is that people may change the naming of the MLIR attributes in LLVM dialect, and do not run any Flang testing (I think it is a usual practice to limit the testing to the components that your changes affect).  Flang LIT testing will break after the commit, while it is preferable not to have the breakage in the first place.  So I would prefer any solution that breaks the names dependency between Flang and LLVM dialect. Using the getters from LLVMFuncOp sounds acceptable. Moving this pass after the `FIRToLLVMLowering` pass seems to be an option as well.

asb replied:
@vzakhari Though it's probably common to not build-test flang as well in that kind of scenario. I'm not sure I see any appropriate static getters in LLVMFuncOp that would get this information, and I don't believe I can get an instance of an LLVMFuncOp at this stage (though could be wrong - I'm new to MLIR and Flang). Could you please point to what you were imagining I could use?

vzakhari replied:
After #79812 we should get static methods in `LLVMFuncOp` class that return the required attribute names, e.g. `getUnsafeFpMathAttrName`, `getNoNansFpMathAttrName`, etc.  You may look it up in `tools/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.h.inc` in your llvm build location.

asb committed:
Modify func-attr-fast-math.f90 test case to be resilient to other attributes being produced (e.g. target-cpu, target-features)

asb commented:
Rebased now that the dependent MLIR patch has landed and adjusted test case now that target-cpu and target-features are also produced in emitted IR.

banach-space commented on flang/test/Driver/func-attr-fast-math.f90:
Wouldn’t this test work for any target? Just wondering why limit it to X86?

asb replied:
Thanks - it doesn't. I'd started down the path of checking all attributes, then switched to the better approach of skipping irrelevant ones with a regex but didn't revert the x86-specific bit.

asb committed:
func-attr-fast-math.f90 doesn't need to be x86-specific

asb committed:
Use static LLVMFuncOp helpers to get attribute names

As requested during review.

asb commented:
I believe all review comments are now addressed.

vzakhari commented on flang/lib/Optimizer/Transforms/FunctionAttr.cpp:
```diff
@@ -45,14 +51,42 @@ void FunctionAttrPass::runOnOperation() {
     func->setAttr("frame_pointer", mlir::LLVM::FramePointerKindAttr::get(
                                        context, framePointerKind));
 
+  auto llvmFuncOpName = mlir::OperationName("llvm.func", context);
```
```suggestion
  auto llvmFuncOpName = mlir::OperationName(LLVMFuncOp::getOperationName(), context);
```

vzakhari reviewed (approved):
Thank you! Looks good (with one minor suggestions inlined).

asb committed:
Typo fix, and apply suggestion from review
